src/rgw/rgw_d3n_cacherequest.cc

```diff
@@ -0,0 +1,146 @@
+#include "common/errno.h"
+
+
+namespace rgw::d3n {
+
+// unique_ptr with custom deleter for struct aiocb
+struct libaio_aiocb_deleter {
+  void operator()(struct aiocb* c) {
+    if(c->aio_fildes > 0) {
+      if( ::close(c->aio_fildes) != 0) {
+        lsubdout(g_ceph_context, rgw_datacache, 2) << "D3nDataCache: " << __func__ << "(): Error - can't close file, errno=" << -errno << dendl;
+      }
+    }
+    delete c;
+  }
+};
+
+using unique_aio_cb_ptr = std::unique_ptr<struct aiocb, libaio_aiocb_deleter>;
+
+using aio_notify_fn = void(*)(union sigval); // sigevent::sigev_notify_function
+
+struct AsyncFileReadOp {
+  bufferlist result;
+  unique_aio_cb_ptr aio_cb;
+
+  int init_async_read(const DoutPrefixProvider *dpp, const std::string& location,
+                      off_t read_ofs, off_t read_len, aio_notify_fn notify_cb,
+                      void* arg)
+  {
+    ldpp_dout(dpp, 20) << "D3nDataCache: " << __func__ << "(): location=" << location << dendl;
+    aio_cb.reset(new struct aiocb);
+    memset(aio_cb.get(), 0, sizeof(struct aiocb));
+    aio_cb->aio_fildes = TEMP_FAILURE_RETRY(::open(location.c_str(), O_RDONLY|O_CLOEXEC|O_BINARY));
+    if(aio_cb->aio_fildes < 0) {
+      int err = errno;
+      ldpp_dout(dpp, 1) << "ERROR: D3nDataCache: " << __func__ << "(): can't open " << location << " : " << cpp_strerror(err) << dendl;
+      return -err;
+    }
+    if (g_conf()->rgw_d3n_l1_fadvise != POSIX_FADV_NORMAL)
+      posix_fadvise(aio_cb->aio_fildes, 0, 0, g_conf()->rgw_d3n_l1_fadvise);
+
+    bufferptr bp(read_len);
+    aio_cb->aio_buf = bp.c_str();
+    result.append(std::move(bp));
+
+    aio_cb->aio_nbytes = read_len;
+    aio_cb->aio_offset = read_ofs;
+    aio_cb->aio_sigevent.sigev_notify = SIGEV_THREAD;
+    aio_cb->aio_sigevent.sigev_notify_function = notify_cb;
+    aio_cb->aio_sigevent.sigev_notify_attributes = nullptr;
+    aio_cb->aio_sigevent.sigev_value.sival_ptr = arg;
+
+    return 0;
+  }
+};
+
+using Signature = void(boost::system::error_code, bufferlist);
+using Completion = ceph::async::Completion<Signature, AsyncFileReadOp>;
+
+static void on_notify_yield(sigval sigval)
+{
+  lsubdout(g_ceph_context, rgw_datacache, 20) << "D3nDataCache: " << __func__ << "()" << dendl;
+  auto p = std::unique_ptr<Completion>{static_cast<Completion*>(sigval.sival_ptr)};
+  auto op = std::move(p->user_data);
+  const int ret = -aio_error(op.aio_cb.get());
+  boost::system::error_code ec;
+  if (ret < 0) {
+    ec.assign(-ret, boost::system::system_category());
+  }
+
+  ceph::async::dispatch(std::move(p), ec, std::move(op.result));
+}
+
+template <typename ExecutionContext, typename CompletionToken>
+static auto async_read(const DoutPrefixProvider *dpp, ExecutionContext& ctx,
+                       const std::string& location, off_t ofs, off_t len,
+                       CompletionToken&& token)
+{
+  boost::asio::async_completion<CompletionToken, Signature> init(token);
+  auto p = Completion::create(ctx.get_executor(),
```

Comment:
@cbodley : As you said in the stand up - the strand here is the execution context. I have additional questions here:
1. Why is the io_context then being used as the executor in this Completion::create() call? 
2. I understand that a strand is used to serialize handlers - are we going to serialize the execution of d3n_ibaio_handler(), as a result of ceph::async::post/ ceph::async::dispatch.
3. Since async_read() is being invoked as a result of Throttle::get(), is this getting executed as part of a coroutine? I dont think it is - correct me if I am wrong.

Reply:
> 1. Why is the io_context then being used as the executor in this Completion::create() call?

the real `init.completion_handler` passed below may not always have an 'associated executor'; when it doesn't, `Completion::create()` would use `ctx.get_executor()` by default. but in `cache_read()` below, we use `bind_executor()` to make sure that the `yield_context`'s executor is the handler's 'associated executor':
```c++
  async_completion<yield_context, void()> init(yield);
  auto ex = get_associated_executor(init.completion_handler);

  async_read(dpp, context, path, ofs, len, r, buffer,
             bind_executor(ex, d3n_libaio_handler{r}));
```
so `ctx.get_executor()` isn't used to invoke the `d3n_libaio_handler` callback, but it would be if we omitted the `bind_executor()` part

another responsibility of `Completion` is to maintain 'work' on these execution contexts. `io_context::run()` will keep going until all work is finished. maintaining work on both executors means that they'll continue waiting until we post/dispatch that completion at the end. in this `yield_context` case, we do know that the associated executor is a strand that wraps the same execution context `ctx`, but in the general case they're allowed to be different

Reply:
> 3. Since async_read() is being invoked as a result of Throttle::get(), is this getting executed as part of a coroutine? does init.result.get(), suspend the coroutine() and then ceph::async::dispatch() resume it?

regarding suspend/resume, that's essentially how it works when you use `yield_context` directly as the `CompletionToken` argument

but the interaction between these aio ops and the `AioThrottle` class is a bit more subtle, because we're trying to run several of these ops in parallel with the request coroutine. `YieldingAioThrottle` is the thing that will actually suspend the coroutine, either when `YieldingAioThrottle::get()` would exceed the throttle window, or when `YieldingAioThrottle::drain()` wants to wait for all of the completions. each individual aio op just calls `AioThrottle::put()` on completion - that may or may not wake up the calling coroutine, depending on whether/why it was blocked in the first place

with `BlockingAioThrottle`, we get the completions from a background librados thread, which is different from the frontend thread handling the request. so we need a mutex/condition_variable to make `BlockingAioThrottle::put()` thread-safe

with `YieldingAioThrottle`, we just have to make sure that the calls to `YieldingAioThrottle::put()` happen from within that strand executor. that's why `cache_read()` uses `get_associated_executor()`/`bind_executor()` instead of passing the `yield_context` directly

Reply:
@cbodley: so this means, there is no supension/ resume happening in `init.result.get()/ceph::async::dispatch()`. That is happening as part of `YieldingAioThrottle::get/put()`. In d3n cache code, work is maintained on the execution contexts by `Completion`, till `libaiohandler` s invoked. - Is my understanding correct now?

Reply:
> Is my understanding correct now?

that's right